net/asymcute: fix MsgId when publishing with QOS0

haukepetersen opened the pull request:
### Contribution description
The MQTT-SN standard states in 5.4.12 for `PUBLISH` messages: "MsgId: same meaning as the MQTT “Message ID”; only relevant in case of QoS levels 1 and 2, otherwise coded 0x0000.". So far we do this wrong and set the MsgId to an incrementing value every time we send a `PUBLISH` message. This PR fixes that...

I found this when using the Paho MQTT-SN gateway (https://github.com/eclipse/paho.mqtt-sn.embedded-c/tree/master/MQTTSNGateway), where the gateway did not want to forward my PUBLISH messages with the reason of `Invalid MsgId`...

### Testing procedure
Run the `examples/asymcute` example application on native and observe the tap interface using wirshark. When looking into the payload of publish messages, the MsgId field should be set to `0` when sending QOS0, and to some number when sending QOS1.

### Issues/PRs references
none

miri64 commented:
Do we also need to fix that in `emcute`?

haukepetersen commented:
Possibly?! Will check...

haukepetersen commented:
fixed there as well: #15661

miri64 commented on sys/net/application_layer/asymcute/asymcute.c:
```diff
@@ -879,8 +879,8 @@ int asymcute_publish(asymcute_con_t *con, asymcute_req_t *req,
-    /* get message id */
-    req->msg_id = _msg_id_next(con);
+    /* get message id, set to 0 for QOS 0 */
+    req->msg_id = (flags & MQTTSN_QOS_MASK) ? _msg_id_next(con) : 0;
```
From OP:

> The MQTT-SN standard states in 5.4.12 for PUBLISH messages: "MsgId: same meaning as the MQTT “Message ID”; only relevant in case of QoS levels 1 and 2, otherwise coded 0x0000.". 

So it must also be 0 for QoS -1 (_negative 1_), right? So there also needs to be a check if both QoS flags are set.

haukepetersen replied:
Yes. did forget about QOS -1. Although it is not really applicable for the current implementation, as only QOS1 is supported and hence `MQTTSN_QOS_0` and `MQTTSN_QOS_1` are the only two flags currently allowed in that part of the flags field... But does not hurt to be more specific here in any case, will fix.

haukepetersen commented:
addressed comment:
- set msgId to 0 for both QoS0 and QoS -1, despite QoS -1 not being allowed at the moment...
- add doc about QoS -1 not being supported :-)
- add define for the QoS -1 flag to `mqttsn.h`

miri64 reviewed (approved):
Can't test this as proposed as the broker won't compile on arch linux at the moment for me

<details>

```
/usr/bin/ld: /tmp/ccIwbXrR.o:(.bss+0x8): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccm4YwRQ.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccd9Sw0Q.o:(.bss+0x20): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccMUfM3Q.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/cctndH9S.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/cc063v5Q.o:(.bss+0x60): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccZOqw6Q.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccBGooNT.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccEBembQ.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccc4Rp2T.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/cclufiMR.o:(.bss+0x4): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccVljqFS.o:(.bss+0x0): multiple definition of `bstate'; /tmp/cc063v5Q.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccVljqFS.o:(.bss+0x8): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccm4lJsT.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccte83vR.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccWkBfHS.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/ccUJ2v6R.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/cc6MVdWP.o:(.bss+0x0): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/cc14eesP.o:(.bss+0x0): multiple definition of `bstate'; /tmp/cc063v5Q.o:(.bss+0x0): first defined here
/usr/bin/ld: /tmp/cc14eesP.o:(.bss+0x8): multiple definition of `Log_levels'; /tmp/ccbYemFR.o:(.bss+0x0): first defined here
```

</details>

However, the changes look sane and simple enough. ACK!

miri64 commented:
Please squash.

haukepetersen commented:
I adapted the condition statement once more, as IMHO this way it is much easier to read and it uses the same semantic than the fix in #15661. Changes are squashed.

miri64 commented:
ACK upholds.
